# Needed modules

In [5]:
import numpy as np
import pandas as pd
import os
import pickle
import sqlite3

In [6]:
# https://kioku-space.com/en/jupyter-skip-execution/
from IPython.core.magic import register_cell_magic # type: ignore

@register_cell_magic
def skip(line, cell):
    return

# Pickle save

In [7]:
%%skip
# =============================================================================
# main function
# =============================================================================
def data_save_load(option):
    """
    This function is used to save or load data for the jupyter notebook
    """
    path_folder = "ipynb_db"  # Folder to save variables
    os.makedirs(path_folder, exist_ok=True)  # Create folder if not exist
    path_file = os.path.join(path_folder, "variables.pkl") # Path to save the variables

    if option == "save":
        # Save the variables
        variables_dict = {
            "df_main" : df_main
            }

        with open(path_file, "wb") as f:
            pickle.dump(variables_dict, f)


    elif option == "load":
        with open(path_file, "rb") as f:
            variables = pickle.load(f)
        # Now load the variables
        for key, value in variables.items():
            print (f"Loading variable: {key}") 
            globals()[key] = value  # Assign the value to the variable

# =============================================================================
# Call the function
# =============================================================================
data_save_load(option="save")

# 1. Preparing data

Bear in mind the original wikidata table has like 36 tabs in the excel file that are of our interest. Apart from that 36 files there are some that we don't want, such as "VARIOS" or "C-Sastre"

## 1.1 Load data

In [8]:
# Path to the data
path_wiki_xlsx = "~/Downloads/tmp/20240723142835_LinJPC-Wikipedia-2024.xlsx"

In [9]:
# Needed columns in the excel file
needed_sheets = [f"Crom{i}" for i in range (1, 36+1)]
print(needed_sheets) 

['Crom1', 'Crom2', 'Crom3', 'Crom4', 'Crom5', 'Crom6', 'Crom7', 'Crom8', 'Crom9', 'Crom10', 'Crom11', 'Crom12', 'Crom13', 'Crom14', 'Crom15', 'Crom16', 'Crom17', 'Crom18', 'Crom19', 'Crom20', 'Crom21', 'Crom22', 'Crom23', 'Crom24', 'Crom25', 'Crom26', 'Crom27', 'Crom28', 'Crom29', 'Crom30', 'Crom31', 'Crom32', 'Crom33', 'Crom34', 'Crom35', 'Crom36']


In [10]:
# Read the data
pre_wiki_df = pd.read_excel(path_wiki_xlsx, 
                            sheet_name=needed_sheets,
                            engine='openpyxl')

In [11]:
# Merge all Data Frames into a single one
wiki_df = pd.concat(pre_wiki_df.values(), ignore_index=True, sort=False)

# Replace missing values with NaN
wiki_df = wiki_df.fillna(value=np.nan)

In [12]:
# Inspect data frame
print(wiki_df.shape)
print(wiki_df.dtypes)
wiki_df.head()

(8881, 30)
Etiqueta                  object
Wikidata                  object
Mendeley dataset          object
Description               object
Descripción               object
Otros nombres             object
LMJFC ortólogo            object
LmjF ortólogo             object
Función molecular         object
Referencia1               object
Referencia2               object
Otras Referencias         object
Gen                       object
CDS                       object
Proteina                  object
Notas                     object
Proteoma Experimental     object
LBRM2904 ortólogo         object
Wikidata-LbrM2904         object
LdHU3 ortólogo            object
Otras referencias         object
Proteína                  object
Gene                      object
OtrasReferencias          object
LbrM2904 ortólogo         object
LbrM ortólogo             object
Unnamed: 16              float64
Aminoácidos               object
Imagen                    object
Imagen Wiki               object

Etiqueta   Wikidata                         Mendeley dataset  \
0  LINF_010005000  Q62005644  http://dx.doi.org/10.17632/m9mggnd5w4.1   
1  LINF_010005100  Q65246947  http://dx.doi.org/10.17632/d99ycn2bhb.1   
2  LINF_010005200  Q65459375  http://dx.doi.org/10.17632/vsmrprnpb5.2   
3  LINF_010005300  Q65461401  http://dx.doi.org/10.17632/78btgwdkx9.1   
4  LINF_010005400  Q65463295  http://dx.doi.org/10.17632/b79hrctkww.1   

                                   Description  \
0        Protein of unknown function (DUF2946)   
1  Endonuclease/Exonuclease/phosphatase family   
2                                   Kinesin-13   
3             hypothetical protein - conserved   
4        Propionyl-CoA carboxylase - α-subunit   

                                         Descripción Otros nombres  \
0  Proteína de función desconocida con dominio DU...  LinJ.01.0010   
1  proteína de la familia de endonucleasa/exonucl...  LinJ.01.0020   
2                                       Kinesina-13   LinJ.01.0030   
3                   Proteína hipotética - conservada  LinJ.01.0040   
4            Propionil-CoA carboxilasa - subunidad α  LinJ.01.0050   

    LMJFC ortólogo LmjF ortólogo Función molecular  \
0  LMJFC_010005100  LmjF.01.0010               NaN   
1  LMJFC_010005200  LmjF.01.0020               NaN   
2  LMJFC_010005300  LmjF.01.0030               NaN   
3  LMJFC_010005400  LmjF.01.0040               NaN   
4  LMJFC_010005500  LmjF.01.0050               NaN   

                                         Referencia1  ... Otras referencias  \
0                                                NaN  ...               NaN   
1                                                NaN  ...               NaN   
2                                                NaN  ...               NaN   
3                                                NaN  ...               NaN   
4  Lee, J.K.J., Liu, Y.T., Hu, J.J., Aphasizheva,...  ...               NaN   

  Proteína Gene OtrasReferencias LbrM2904 ortólogo LbrM ortólogo Unnamed: 16  \
0      NaN  NaN              NaN               NaN           NaN         NaN   
1      NaN  NaN              NaN               NaN           NaN         NaN   
2      NaN  NaN              NaN               NaN           NaN         NaN   
3      NaN  NaN              NaN               NaN           NaN         NaN   
4      NaN  NaN              NaN               NaN           NaN         NaN   

  Aminoácidos Imagen Imagen Wiki  
0         NaN    NaN         NaN  
1         NaN    NaN         NaN  
2         NaN    NaN         NaN  
3         NaN    NaN         NaN  
4         NaN    NaN         NaN  

[5 rows x 30 columns]

## 1.2 Correct data

In [13]:
# Let's check again only the columns
wiki_df.dtypes

Etiqueta                  object
Wikidata                  object
Mendeley dataset          object
Description               object
Descripción               object
Otros nombres             object
LMJFC ortólogo            object
LmjF ortólogo             object
Función molecular         object
Referencia1               object
Referencia2               object
Otras Referencias         object
Gen                       object
CDS                       object
Proteina                  object
Notas                     object
Proteoma Experimental     object
LBRM2904 ortólogo         object
Wikidata-LbrM2904         object
LdHU3 ortólogo            object
Otras referencias         object
Proteína                  object
Gene                      object
OtrasReferencias          object
LbrM2904 ortólogo         object
LbrM ortólogo             object
Unnamed: 16              float64
Aminoácidos               object
Imagen                    object
Imagen Wiki               object
dtype: obj

The only columns needed are the next ones:
* **Etiqueta**: Gene ID.
* **Description**: names to use for the element. If there are more than one, it's separated by "|".
* **Otros nombres**: old names when they were discovered first.
* **Orthologs**: different orthologs in *L. major*, *L. donovani*, *L. braziliensis*.
  * **LMJFC ortólogo**: old *L. major* strand ortholog.
  * **LmjF ortólogo**: current *L. major* strand ortholog.
  * **LBRM2904 ortólogo**: *L. braziliensis* ortholog? <mark>CHECK IT</mark>
  * **LdHU3 ortólogo**: *L. donovani* ortholog.
  * **LbrM2904 ortólogo**: *L. braziliensis* ortholog.
  * **LbrM ortólogo**: *L. braziliensis* ortholog.
* **Wikidata**: ID to search in wikidata.
* **Mendeley dataset**: ID for Mendeley data.  

Apart from that, we need another column for **Uniprot** searching.

We can see that we need a lot of columns, But we shall be careful with:
* **Orthologs**: there are a lot of them and I don't know if all of them are useful.
* **Mendeley data**: two versions or what?

### 1.2.1 Take needed columns

In [14]:
needed_cols = [
    "Etiqueta",
    "Description",
    "Otros nombres",
    "Wikidata",
    "Mendeley dataset",
    "LmjF ortólogo",
    "LMJFC ortólogo",
    "LdHU3 ortólogo",
    "LBRM2904 ortólogo",
    "LbrM2904 ortólogo",
    "LbrM ortólogo"
    ]

df_main = wiki_df[needed_cols]
print(df_main.shape)
df_main.head()

(8881, 11)


Etiqueta                                  Description Otros nombres  \
0  LINF_010005000        Protein of unknown function (DUF2946)  LinJ.01.0010   
1  LINF_010005100  Endonuclease/Exonuclease/phosphatase family  LinJ.01.0020   
2  LINF_010005200                                   Kinesin-13  LinJ.01.0030   
3  LINF_010005300             hypothetical protein - conserved  LinJ.01.0040   
4  LINF_010005400        Propionyl-CoA carboxylase - α-subunit  LinJ.01.0050   

    Wikidata                         Mendeley dataset LmjF ortólogo  \
0  Q62005644  http://dx.doi.org/10.17632/m9mggnd5w4.1  LmjF.01.0010   
1  Q65246947  http://dx.doi.org/10.17632/d99ycn2bhb.1  LmjF.01.0020   
2  Q65459375  http://dx.doi.org/10.17632/vsmrprnpb5.2  LmjF.01.0030   
3  Q65461401  http://dx.doi.org/10.17632/78btgwdkx9.1  LmjF.01.0040   
4  Q65463295  http://dx.doi.org/10.17632/b79hrctkww.1  LmjF.01.0050   

    LMJFC ortólogo LdHU3 ortólogo LBRM2904 ortólogo LbrM2904 ortólogo  \
0  LMJFC_010005100  LDHU3_01.0030               NaN               NaN   
1  LMJFC_010005200  LDHU3_01.0040  LBRM2904_01.0010               NaN   
2  LMJFC_010005300  LDHU3_01.0050  LBRM2904_01.0020               NaN   
3  LMJFC_010005400  LDHU3_01.0060  LBRM2904_01.0030               NaN   
4  LMJFC_010005500  LDHU3_01.0070  LBRM2904_01.0040               NaN   

  LbrM ortólogo  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN

### 1.2.2 Check the *L. braziliensis* orthologs

Let's check all the conflicted columns:

In [15]:
# Columns to observe:
braz_ortho_cols = [
    "LBRM2904 ortólogo",
    "LbrM2904 ortólogo",
    "LbrM ortólogo"
]

In [16]:
# Check the values:
for i in range(len(braz_ortho_cols)):
    mask = braz_ortho_cols[i]
    print(f"Numpy NaN values: {df_main[mask].isna().sum()}")
    print(df_main[mask].value_counts())
    if i != len(braz_ortho_cols) - 1:
        print("\n", "="*50, "\n", sep="")

Numpy NaN values: 8613
LBRM2904 ortólogo
-                    3
LBRM2904_06.0090     2
LBRM2904_05.0520     2
LBRM2904_05.0200     1
LBRM2904_05.0160     1
                    ..
LBRM2904_02.0140     1
LBRM2904_02.0070     1
LBRM2904_02.0030     1
LBRM2904_02.0010     1
LBRM2904_02.0390     1
Name: count, Length: 264, dtype: int64


Numpy NaN values: 8879
LbrM2904 ortólogo
LBRM2904_13.1180|LBRM2904_13.1190    1
LBRM2904_28.2560                     1
Name: count, dtype: int64


Numpy NaN values: 8875
LbrM ortólogo
LBRM2904_17.1610    1
LBRM2904_25.0530    1
LBRM2904_30.2040    1
LBRM2904_30.3180    1
LBRM2904_31.2420    1
LBRM2904_20.0020    1
Name: count, dtype: int64


We can observe the following:
* **LBRM2904 ortólogo**: has 8613 NaN values (the less of the three) so this one es the main one to use.
* **LbrM2904 ortólogo**: except for 2 values, everything is NaN.
* **LbrM ortólogo**: except for 6 values, everything is NaN.

In [17]:
# Check where "LbrM2904 ortólogo* is not NaN
df_main.loc[
    df_main[braz_ortho_cols[1]].notna(),
    braz_ortho_cols
    ]

LBRM2904 ortólogo                  LbrM2904 ortólogo LbrM ortólogo
1696               NaN  LBRM2904_13.1180|LBRM2904_13.1190           NaN
5069               NaN                   LBRM2904_28.2560           NaN

We can see how **LBRM2904 ortólogo** is NaN here. I the data of **LbrM2904 ortólogo** should be in **LBRM2904 ortólogo**

In [18]:
# Check where "LbrM ortólogo" is not NaN
df_main.loc[
    df_main[braz_ortho_cols[2]].notna(),
    braz_ortho_cols
    ]

LBRM2904 ortólogo LbrM2904 ortólogo     LbrM ortólogo
2480               NaN               NaN  LBRM2904_17.1610
4016               NaN               NaN  LBRM2904_25.0530
5681               NaN               NaN  LBRM2904_30.2040
5794               NaN               NaN  LBRM2904_30.3180
6156               NaN               NaN  LBRM2904_31.2420
7063               NaN               NaN  LBRM2904_20.0020

My guess is the same as before, all this data should be in **LBRM2904 ortólogo**

In [19]:
# Let's merge the two columns into "LBRM2904 ortólogo"
# First with the "LbrM2904 ortólogo" column
df_main.loc[:,"LBRM2904 ortólogo"] = df_main["LBRM2904 ortólogo"].combine_first(df_main["LbrM2904 ortólogo"])

# Now with the "LbrM ortólogo" column
df_main.loc[:,"LBRM2904 ortólogo"] = df_main["LBRM2904 ortólogo"].combine_first(df_main["LbrM ortólogo"])

In [20]:
# Let's check if it worked
df_main.loc[df_main[braz_ortho_cols[1]].notna(), braz_ortho_cols]

LBRM2904 ortólogo                  LbrM2904 ortólogo  \
1696  LBRM2904_13.1180|LBRM2904_13.1190  LBRM2904_13.1180|LBRM2904_13.1190   
5069                   LBRM2904_28.2560                   LBRM2904_28.2560   

     LbrM ortólogo  
1696           NaN  
5069           NaN

In [21]:
# Check the other one
df_main.loc[df_main[braz_ortho_cols[2]].notna(), braz_ortho_cols]

LBRM2904 ortólogo LbrM2904 ortólogo     LbrM ortólogo
2480  LBRM2904_17.1610               NaN  LBRM2904_17.1610
4016  LBRM2904_25.0530               NaN  LBRM2904_25.0530
5681  LBRM2904_30.2040               NaN  LBRM2904_30.2040
5794  LBRM2904_30.3180               NaN  LBRM2904_30.3180
6156  LBRM2904_31.2420               NaN  LBRM2904_31.2420
7063  LBRM2904_20.0020               NaN  LBRM2904_20.0020

Since it worked, let's remove the columns **LbrM2904 ortólo** and **LbrM ortólogo**

In [22]:
df_main = df_main.drop(columns=braz_ortho_cols[1:])
df_main.head()

Etiqueta                                  Description Otros nombres  \
0  LINF_010005000        Protein of unknown function (DUF2946)  LinJ.01.0010   
1  LINF_010005100  Endonuclease/Exonuclease/phosphatase family  LinJ.01.0020   
2  LINF_010005200                                   Kinesin-13  LinJ.01.0030   
3  LINF_010005300             hypothetical protein - conserved  LinJ.01.0040   
4  LINF_010005400        Propionyl-CoA carboxylase - α-subunit  LinJ.01.0050   

    Wikidata                         Mendeley dataset LmjF ortólogo  \
0  Q62005644  http://dx.doi.org/10.17632/m9mggnd5w4.1  LmjF.01.0010   
1  Q65246947  http://dx.doi.org/10.17632/d99ycn2bhb.1  LmjF.01.0020   
2  Q65459375  http://dx.doi.org/10.17632/vsmrprnpb5.2  LmjF.01.0030   
3  Q65461401  http://dx.doi.org/10.17632/78btgwdkx9.1  LmjF.01.0040   
4  Q65463295  http://dx.doi.org/10.17632/b79hrctkww.1  LmjF.01.0050   

    LMJFC ortólogo LdHU3 ortólogo LBRM2904 ortólogo  
0  LMJFC_010005100  LDHU3_01.0030               NaN  
1  LMJFC_010005200  LDHU3_01.0040  LBRM2904_01.0010  
2  LMJFC_010005300  LDHU3_01.0050  LBRM2904_01.0020  
3  LMJFC_010005400  LDHU3_01.0060  LBRM2904_01.0030  
4  LMJFC_010005500  LDHU3_01.0070  LBRM2904_01.0040

### 1.2.3 Check for np.nan values in "Etiqueta"

In [23]:
# Checking np.nan values in the "Etiqueta" column
print(f"Numpy NaN values in 'Etiqueta' column: {df_main['Etiqueta'].isna().sum()}")
etiqueta_nan = df_main.loc[df_main['Etiqueta'].isna()]
etiqueta_nan

Numpy NaN values in 'Etiqueta' column: 155


Etiqueta Description Otros nombres Wikidata Mendeley dataset  \
161       NaN         NaN           NaN      NaN              NaN   
162       NaN         NaN           NaN      NaN              NaN   
163       NaN         NaN           NaN      NaN              NaN   
164       NaN         NaN           NaN      NaN              NaN   
264       NaN         NaN           NaN      NaN              NaN   
...       ...         ...           ...      ...              ...   
8098      NaN         NaN           NaN      NaN              NaN   
8877      NaN         NaN           NaN      NaN              NaN   
8878      NaN         NaN           NaN      NaN              NaN   
8879      NaN         NaN           NaN      NaN              NaN   
8880      NaN         NaN           NaN      NaN              NaN   

     LmjF ortólogo LMJFC ortólogo LdHU3 ortólogo LBRM2904 ortólogo  
161            NaN            NaN            NaN               NaN  
162            NaN            NaN            NaN               NaN  
163            NaN            NaN            NaN               NaN  
164            NaN            NaN            NaN               NaN  
264            NaN            NaN            NaN               NaN  
...            ...            ...            ...               ...  
8098           NaN            NaN            NaN               NaN  
8877           NaN            NaN            NaN               NaN  
8878           NaN            NaN            NaN               NaN  
8879           NaN            NaN            NaN               NaN  
8880           NaN            NaN            NaN               NaN  

[155 rows x 9 columns]

In [144]:
# Remove those values from the df_main
df_main = df_main.dropna(subset=["Etiqueta"])  # Remove the rows with np.nan values in the "Etiqueta" column

# Check again
print(f"Numpy NaN values in 'Etiqueta' column: {df_main['Etiqueta'].isna().sum()}")

Numpy NaN values in 'Etiqueta' column: 0


### 1.2.4 Check for elements that don't start with "LINF"

In [30]:
# Original shape of df_main
print(df_main.shape)

(8881, 9)


In [28]:
# Check the elements in column 'Etiqueta' that don't start with "LINF"
df_main[df_main["Etiqueta"].str.contains(r"^(?!LINF)", na=False)]

Etiqueta           Description  \
710               CDS embebida en CDS de LINF_070008000                   NaN   
2081                                        No ortólogo                   NaN   
2082                                        No ortólogo                   NaN   
4746  No parece existir en L. infantum, sí que está ...                   NaN   
4915                                        No ortólogo  hypothetical protein   
5623                          No parece existir en LINF                   NaN   
5626                          No parece existir en LINF                   NaN   

     Otros nombres Wikidata Mendeley dataset                    LmjF ortólogo  \
710            NaN      NaN              NaN                              NaN   
2081           NaN      NaN              NaN                     LmjF.15.1120   
2082           NaN      NaN              NaN                     LmjF.15.1160   
4746           NaN      NaN              NaN                    LmjF.27.T2035   
4915           NaN      NaN              NaN                    LmjF.28.T0985   
5623           NaN      NaN              NaN  Posible proteína: ATG en 489011   
5626           NaN      NaN              NaN  Posible proteína: ATG en 495185   

     LMJFC ortólogo LdHU3 ortólogo LBRM2904 ortólogo  
710             NaN            NaN               NaN  
2081            NaN            NaN               NaN  
2082            NaN            NaN               NaN  
4746            NaN            NaN               NaN  
4915            NaN            NaN               NaN  
5623            NaN            NaN               NaN  
5626            NaN            NaN               NaN

In [34]:
# Remove those values from the df_main 
df_main = df_main[~df_main["Etiqueta"].str.contains(r"^(?!LINF)", na=False)]
# Now restart the index
df_main = df_main.reset_index(drop=True)
print(df_main.shape)

(8874, 9)


## 1.3 Modify data

### 1.3.1 Description column

Here are the names, something split by "|", let's split them by ","

In [145]:
# Check elements with "|"; be careful with the numpy NaN values
df_main.loc[df_main["Description"].str.contains(r"\|", na=False), "Description"]

9                                 Fucosyltransferase|FUT2
20                                            ZC3H27|CSBP
21                                   cyclophilin 12|CyP12
27      mt-LAF3|mitoribosomal LSU assembly factor 3|ps...
34                    Mitochondrial pyruvate carrier|MPC1
                              ...                        
8865              RNA editing complex protein MP63|KREPA2
8867                  Protein disulfide isomerase 2|PDI52
8871    eukaryotic translation initiation factor 3 sub...
8874                       Histone acetyltransferase|HAT3
8875                  Cytochrome oxidase subunit IX|COXIX
Name: Description, Length: 1573, dtype: object

In [146]:
# Let's read the "Description" column and if there are values separated by "|", let's split them by "," instead
df_main.loc[:, "Description"] = df_main["Description"].str.replace("|", ";")

In [147]:
# Check if it worked by searching "|" again
df_main.loc[df_main["Description"].str.contains(r"\|", na=False), "Description"]

Series([], Name: Description, dtype: object)

In [148]:
# And searching ";" now:
df_main.loc[df_main["Description"].str.contains(r";", na=False), "Description"]

9                                 Fucosyltransferase;FUT2
20                                            ZC3H27;CSBP
21                                   cyclophilin 12;CyP12
27      mt-LAF3;mitoribosomal LSU assembly factor 3;ps...
34                    Mitochondrial pyruvate carrier;MPC1
                              ...                        
8865              RNA editing complex protein MP63;KREPA2
8867                  Protein disulfide isomerase 2;PDI52
8871    eukaryotic translation initiation factor 3 sub...
8874                       Histone acetyltransferase;HAT3
8875                  Cytochrome oxidase subunit IX;COXIX
Name: Description, Length: 1573, dtype: object

### 1.3.2 Wikidata column

For example, to search for a the wiki data "Q62005644" it would be something like: https://www.wikidata.org/wiki/Q62005644

In [149]:
# Check if there are numpy nan values in the "WikiData" column
df_main["Wikidata"].isna().sum()

np.int64(3218)

In the np.nan cases, let it stay as np.nan

In [150]:
# In the "Wikidata" column, if there are not NaN values, let's create a URL like https://www.wikidata.org/wiki/x where x is the value in the "Wikidata" column. 
df_main.loc[df_main["Wikidata"].notna(), "Wikidata"] = "https://www.wikidata.org/wiki/" + df_main.loc[df_main["Wikidata"].notna(), "Wikidata"]

# Check if it worked
df_main.loc[df_main["Wikidata"].notna(), "Wikidata"]

0        https://www.wikidata.org/wiki/Q62005644
1        https://www.wikidata.org/wiki/Q65246947
2        https://www.wikidata.org/wiki/Q65459375
3        https://www.wikidata.org/wiki/Q65461401
4        https://www.wikidata.org/wiki/Q65463295
                          ...                   
8871     https://www.wikidata.org/wiki/Q64624105
8873    https://www.wikidata.org/wiki/Q108084512
8874    https://www.wikidata.org/wiki/Q110559335
8875     https://www.wikidata.org/wiki/Q94697040
8876     https://www.wikidata.org/wiki/Q96250673
Name: Wikidata, Length: 5508, dtype: object

In [151]:
# And check the NaN values
df_main.loc[df_main["Wikidata"].isna(), "Wikidata"]

203     NaN
204     NaN
207     NaN
211     NaN
212     NaN
       ... 
8859    NaN
8862    NaN
8866    NaN
8870    NaN
8872    NaN
Name: Wikidata, Length: 3218, dtype: object

### 1.3.3 Mendeley column

In [152]:
# Let's analyze the column first
# Check np.nan values
print(f"NaN values: {df_main["Mendeley dataset"].isna().sum()}")

# Check the values
df_main["Mendeley dataset"].value_counts()

NaN values: 3499


Mendeley dataset
DOI: 10.17632/mjdvj8yrh2.1                 2
DOI:10.17632/9wp56jkpnr.1                  1
http://dx.doi.org/10.17632/g74tpkh8ss.1    1
DOI: 10.17632/m8jvzmw82r.1                 1
DOI:10.17632/dz8wb5z33r.1                  1
                                          ..
DOI:10.17632/zhk2dyygz2.1                  1
DOI:10.17632/59c4jpj6gj.1                  1
DOI:10.17632/67h3b3xzz3.1                  1
DOI:10.17632/cvtbdpwmtj.1                  1
DOI:10.17632/5fcnjt2cvf.1                  1
Name: count, Length: 5226, dtype: int64

We can see there are:
*  **np.nan** values
*  DOI:...
*  http://.....

In [153]:
# Let's take only hte first 4 characters of the "Mendeley dataset" column and then use .value_counts() to see the values
mendeley_col_test = df_main["Mendeley dataset"].str[:4]
mendeley_col_test.value_counts()

Mendeley dataset
DOI:    4542
http     682
doi:       2
10.1       1
Name: count, dtype: int64

Here we know more, there are 4 types:
* DOI:
* http:
* doi:
* 10.1

We'll have to be careful, but what do we need to search in Mendeley dataset?

For example for a DOI such as DOI:10.17632/g74tpkh8ss.1 the web search will be:
* https://data.mendeley.com/datasets/g74tpkh8ss/1
  * We can check how we only need the mixture of letters and numbers after the DOI number. And that the "point" is replaced by "/"

In [154]:
# Let's take the first 14 characters now
mendeley_col_test = df_main["Mendeley dataset"].str[:14]
mendeley_col_test.value_counts()

Mendeley dataset
DOI: 10.17632/    1651
http://dx.doi.     672
DOI:10.17632/5     144
DOI:10.17632/s     134
DOI:10.17632/2     123
DOI:10.17632/p     122
DOI:10.17632/v     122
DOI:10.17632/3     121
DOI:10.17632/b     117
DOI:10.17632/f     115
DOI:10.17632/y     114
DOI:10.17632/9     114
DOI:10.17632/t     114
DOI:10.17632/8     112
DOI:10.17632/z     105
DOI:10.17632/x     104
DOI:10.17632/j     104
DOI:10.17632/r     101
DOI:10.17632/m      99
DOI:10.17632/g      98
DOI:10.17632/n      98
DOI:10.17632/k      97
DOI:10.17632/6      96
DOI:10.17632/w      95
DOI:10.17632/c      92
DOI:10.17632/4      90
DOI:10.17632/h      89
DOI:10.17632/d      87
DOI:10.17632/7      84
https://data.m      10
doi: 10.17632/       2
10.17632/7gc3r       1
Name: count, dtype: int64

Something interesting here is that the DOI number is <"two numbers":"5 numbers"/>. We can use REGEX from the "two numbers":  until the "numenrs"/

In [155]:
# Let's make a regex to extract the numbers after the number like 10.17632/{number}
mendeley_col_test = df_main["Mendeley dataset"].str.extract(r"\d{2}\.\d+/(\S+)")
print(mendeley_col_test.value_counts())
print(type(mendeley_col_test))
print("="*50)

# Replace the "." by "/" in the Data
mendeley_col_test = mendeley_col_test[0].str.replace(".", "/")
print(mendeley_col_test.value_counts())
print("="*50)

# Add at the begining "https://data.mendeley.com/datasets/"
mendeley_col_test = "https://data.mendeley.com/datasets/" + mendeley_col_test
print(mendeley_col_test.value_counts())

# Replace the column df_main["Mendeley dataset"] by the new values in mendeley_col_test
df_main["Mendeley dataset"] = mendeley_col_test

0           
mjdvj8yrh2.1    2
p6c26w5nrf.1    1
p6b9s6zfy5.1    1
p69z3v2b55.1    1
p5pdkzjr3r.1    1
               ..
bsmtxzw63x.1    1
bskf33ws95.1    1
bsk67y5fxb.1    1
bsg3vbyzns.1    1
btjjvsz8z9.1    1
Name: count, Length: 5216, dtype: int64
<class 'pandas.core.frame.DataFrame'>
0
mjdvj8yrh2/1    2
nwtms3s4pj/1    1
p3rpt2dz85/1    1
xhgk4f5wv8/1    1
fvjvwsfjfg/1    1
               ..
pv6bfn472b/1    1
bfn5v8xbbb/1    1
x8p8jpdk25/1    1
498bdcvvkz/1    1
snsdrt6892/1    1
Name: count, Length: 5216, dtype: int64
0
https://data.mendeley.com/datasets/mjdvj8yrh2/1    2
https://data.mendeley.com/datasets/nwtms3s4pj/1    1
https://data.mendeley.com/datasets/p3rpt2dz85/1    1
https://data.mendeley.com/datasets/xhgk4f5wv8/1    1
https://data.mendeley.com/datasets/fvjvwsfjfg/1    1
                                                  ..
https://data.mendeley.com/datasets/pv6bfn472b/1    1
https://data.mendeley.com/datasets/bfn5v8xbbb/1    1
https://data.mendeley.com/datasets/x8p8jpdk25

In [156]:
# Check if it worked
print(df_main["Mendeley dataset"])

# Check head
df_main.head()

0       https://data.mendeley.com/datasets/m9mggnd5w4/1
1       https://data.mendeley.com/datasets/d99ycn2bhb/1
2       https://data.mendeley.com/datasets/vsmrprnpb5/2
3       https://data.mendeley.com/datasets/78btgwdkx9/1
4       https://data.mendeley.com/datasets/b79hrctkww/1
                             ...                       
8872                                                NaN
8873    https://data.mendeley.com/datasets/d9h57y36dc/1
8874    https://data.mendeley.com/datasets/985vk3v8zw/1
8875    https://data.mendeley.com/datasets/mmm3xczkp4/1
8876    https://data.mendeley.com/datasets/hvskp3c57v/1
Name: Mendeley dataset, Length: 8726, dtype: object


Etiqueta                                  Description Otros nombres  \
0  LINF_010005000        Protein of unknown function (DUF2946)  LinJ.01.0010   
1  LINF_010005100  Endonuclease/Exonuclease/phosphatase family  LinJ.01.0020   
2  LINF_010005200                                   Kinesin-13  LinJ.01.0030   
3  LINF_010005300             hypothetical protein - conserved  LinJ.01.0040   
4  LINF_010005400        Propionyl-CoA carboxylase - α-subunit  LinJ.01.0050   

                                  Wikidata  \
0  https://www.wikidata.org/wiki/Q62005644   
1  https://www.wikidata.org/wiki/Q65246947   
2  https://www.wikidata.org/wiki/Q65459375   
3  https://www.wikidata.org/wiki/Q65461401   
4  https://www.wikidata.org/wiki/Q65463295   

                                  Mendeley dataset LmjF ortólogo  \
0  https://data.mendeley.com/datasets/m9mggnd5w4/1  LmjF.01.0010   
1  https://data.mendeley.com/datasets/d99ycn2bhb/1  LmjF.01.0020   
2  https://data.mendeley.com/datasets/vsmrprnpb5/2  LmjF.01.0030   
3  https://data.mendeley.com/datasets/78btgwdkx9/1  LmjF.01.0040   
4  https://data.mendeley.com/datasets/b79hrctkww/1  LmjF.01.0050   

    LMJFC ortólogo LdHU3 ortólogo LBRM2904 ortólogo  
0  LMJFC_010005100  LDHU3_01.0030               NaN  
1  LMJFC_010005200  LDHU3_01.0040  LBRM2904_01.0010  
2  LMJFC_010005300  LDHU3_01.0050  LBRM2904_01.0020  
3  LMJFC_010005400  LDHU3_01.0060  LBRM2904_01.0030  
4  LMJFC_010005500  LDHU3_01.0070  LBRM2904_01.0040

### 1.3.4 Uniprot column

Here we have to search by the unit in UniProt, we would need to use the advance search and search by "Gene Name". The query search would be: `https://www.uniprot.org/uniprotkb?query=%28gene%3A<gene_name>%29`

For example for LINF_010005000 it would be: `ttps://www.uniprot.org/uniprotkb?query=%28gene%3ALINF_010005000%29`

In [157]:
# Create a UniProt Column with the "Etiqueta" column and the url
df_main["UniProt"] = "https://www.uniprot.org/uniprotkb?query=%28gene%3A" + df_main["Etiqueta"] + "%29"

# Move the column after "Mendelet dataset"
col_list = df_main.columns.tolist()  # Get the columns list
col_list.insert(
    col_list.index("Mendeley dataset") + 1,
    col_list.pop(col_list.index("UniProt"))
    )

# Reorder the columns
df_main = df_main[col_list]
print(df_main.columns)
df_main.head()

Index(['Etiqueta', 'Description', 'Otros nombres', 'Wikidata',
       'Mendeley dataset', 'UniProt', 'LmjF ortólogo', 'LMJFC ortólogo',
       'LdHU3 ortólogo', 'LBRM2904 ortólogo'],
      dtype='object')


Etiqueta                                  Description Otros nombres  \
0  LINF_010005000        Protein of unknown function (DUF2946)  LinJ.01.0010   
1  LINF_010005100  Endonuclease/Exonuclease/phosphatase family  LinJ.01.0020   
2  LINF_010005200                                   Kinesin-13  LinJ.01.0030   
3  LINF_010005300             hypothetical protein - conserved  LinJ.01.0040   
4  LINF_010005400        Propionyl-CoA carboxylase - α-subunit  LinJ.01.0050   

                                  Wikidata  \
0  https://www.wikidata.org/wiki/Q62005644   
1  https://www.wikidata.org/wiki/Q65246947   
2  https://www.wikidata.org/wiki/Q65459375   
3  https://www.wikidata.org/wiki/Q65461401   
4  https://www.wikidata.org/wiki/Q65463295   

                                  Mendeley dataset  \
0  https://data.mendeley.com/datasets/m9mggnd5w4/1   
1  https://data.mendeley.com/datasets/d99ycn2bhb/1   
2  https://data.mendeley.com/datasets/vsmrprnpb5/2   
3  https://data.mendeley.com/datasets/78btgwdkx9/1   
4  https://data.mendeley.com/datasets/b79hrctkww/1   

                                             UniProt LmjF ortólogo  \
0  https://www.uniprot.org/uniprotkb?query=%28gen...  LmjF.01.0010   
1  https://www.uniprot.org/uniprotkb?query=%28gen...  LmjF.01.0020   
2  https://www.uniprot.org/uniprotkb?query=%28gen...  LmjF.01.0030   
3  https://www.uniprot.org/uniprotkb?query=%28gen...  LmjF.01.0040   
4  https://www.uniprot.org/uniprotkb?query=%28gen...  LmjF.01.0050   

    LMJFC ortólogo LdHU3 ortólogo LBRM2904 ortólogo  
0  LMJFC_010005100  LDHU3_01.0030               NaN  
1  LMJFC_010005200  LDHU3_01.0040  LBRM2904_01.0010  
2  LMJFC_010005300  LDHU3_01.0050  LBRM2904_01.0020  
3  LMJFC_010005400  LDHU3_01.0060  LBRM2904_01.0030  
4  LMJFC_010005500  LDHU3_01.0070  LBRM2904_01.0040

### 1.3.5 Rename columns

In [158]:
# New names for the columns
new_cols = [
    "Gene_ID",
    "Name",
    "Other_names",
    "Wikidata",
    "Mendeley",
    "UniProt",
    "LmjF_ortholog",
    "LmjFC_ortholog",
    "LdHU3_ortholog",
    "LBRM2904_ortholog"
]

# Rename the columns
df_main.columns = new_cols

# Check the columns
df_main.head()

Gene_ID                                         Name   Other_names  \
0  LINF_010005000        Protein of unknown function (DUF2946)  LinJ.01.0010   
1  LINF_010005100  Endonuclease/Exonuclease/phosphatase family  LinJ.01.0020   
2  LINF_010005200                                   Kinesin-13  LinJ.01.0030   
3  LINF_010005300             hypothetical protein - conserved  LinJ.01.0040   
4  LINF_010005400        Propionyl-CoA carboxylase - α-subunit  LinJ.01.0050   

                                  Wikidata  \
0  https://www.wikidata.org/wiki/Q62005644   
1  https://www.wikidata.org/wiki/Q65246947   
2  https://www.wikidata.org/wiki/Q65459375   
3  https://www.wikidata.org/wiki/Q65461401   
4  https://www.wikidata.org/wiki/Q65463295   

                                          Mendeley  \
0  https://data.mendeley.com/datasets/m9mggnd5w4/1   
1  https://data.mendeley.com/datasets/d99ycn2bhb/1   
2  https://data.mendeley.com/datasets/vsmrprnpb5/2   
3  https://data.mendeley.com/datasets/78btgwdkx9/1   
4  https://data.mendeley.com/datasets/b79hrctkww/1   

                                             UniProt LmjF_ortholog  \
0  https://www.uniprot.org/uniprotkb?query=%28gen...  LmjF.01.0010   
1  https://www.uniprot.org/uniprotkb?query=%28gen...  LmjF.01.0020   
2  https://www.uniprot.org/uniprotkb?query=%28gen...  LmjF.01.0030   
3  https://www.uniprot.org/uniprotkb?query=%28gen...  LmjF.01.0040   
4  https://www.uniprot.org/uniprotkb?query=%28gen...  LmjF.01.0050   

    LmjFC_ortholog LdHU3_ortholog LBRM2904_ortholog  
0  LMJFC_010005100  LDHU3_01.0030               NaN  
1  LMJFC_010005200  LDHU3_01.0040  LBRM2904_01.0010  
2  LMJFC_010005300  LDHU3_01.0050  LBRM2904_01.0020  
3  LMJFC_010005400  LDHU3_01.0060  LBRM2904_01.0030  
4  LMJFC_010005500  LDHU3_01.0070  LBRM2904_01.0040

## 1.4 Prepare data to HTML links

Here we'll need to prepare the columns **Wikidata**, **Mendeley** and **UniProt** so they will have a HTML link and be shorter that way

In [159]:
# Adapt the URL link of the columns Wikidata, Mendeley and UniProt for HTML links "_blank" with the column names as the link name.
# If there are no values, let it stay as NaN
df_main["Wikidata"] = df_main["Wikidata"].apply(lambda x: f'<a href="{x}" target="_blank">Wikidata</a>' if pd.notna(x) else np.nan)
df_main["Mendeley"] = df_main["Mendeley"].apply(lambda x: f'<a href="{x}" target="_blank">Mendeley</a>' if pd.notna(x) else np.nan)
df_main["UniProt"] = df_main["UniProt"].apply(lambda x: f'<a href="{x}" target="_blank">UniProt</a>' if pd.notna(x) else np.nan)

# Check the columns
df_main.loc[:, ["Wikidata", "Mendeley", "UniProt"]].head()

Wikidata  \
0  <a href="https://www.wikidata.org/wiki/Q620056...   
1  <a href="https://www.wikidata.org/wiki/Q652469...   
2  <a href="https://www.wikidata.org/wiki/Q654593...   
3  <a href="https://www.wikidata.org/wiki/Q654614...   
4  <a href="https://www.wikidata.org/wiki/Q654632...   

                                            Mendeley  \
0  <a href="https://data.mendeley.com/datasets/m9...   
1  <a href="https://data.mendeley.com/datasets/d9...   
2  <a href="https://data.mendeley.com/datasets/vs...   
3  <a href="https://data.mendeley.com/datasets/78...   
4  <a href="https://data.mendeley.com/datasets/b7...   

                                             UniProt  
0  <a href="https://www.uniprot.org/uniprotkb?que...  
1  <a href="https://www.uniprot.org/uniprotkb?que...  
2  <a href="https://www.uniprot.org/uniprotkb?que...  
3  <a href="https://www.uniprot.org/uniprotkb?que...  
4  <a href="https://www.uniprot.org/uniprotkb?que...

# 2. SQL database

In [160]:
sqlite_db = "../leishmania.db"

In [161]:
# Connect to the sqlite database
conn = sqlite3.connect(sqlite_db)
cursor = conn.cursor()

# Check if there is already a table claled "infantum"
cursor.execute("SELECT name FROM sqlite_master WHERE type='table' AND name='infantum'")
print(cursor.fetchall())

[('infantum',)]


In [162]:
# Since there already exists a table called "infantum", let's print the schema
cursor.execute("PRAGMA table_info(infantum)")
print(cursor.fetchall())

[(0, 'Gene_ID', 'TEXT', 0, None, 0), (1, 'Name', 'TEXT', 0, None, 0), (2, 'Other_names', 'TEXT', 0, None, 0), (3, 'Wikidata', 'TEXT', 0, None, 0), (4, 'Mendeley', 'TEXT', 0, None, 0), (5, 'UniProt', 'TEXT', 0, None, 0), (6, 'LmjF_ortholog', 'TEXT', 0, None, 0), (7, 'LmjFC_ortholog', 'TEXT', 0, None, 0), (8, 'LdHU3_ortholog', 'TEXT', 0, None, 0), (9, 'LBRM2904_ortholog', 'TEXT', 0, None, 0)]


In [163]:
# Clear the table and replace it with the new data in "col_list" where everythin is a VARCHAR(255) except for the "id" which is an INTEGER PRIMARY KEY and "WikiData", "Mendeley" and "UniProt" which are TEXT
cursor.execute("DROP TABLE infantum")

# Create the table
cursor.execute("""
CREATE TABLE infantum (
    id INTEGER PRIMARY KEY,
    Gene_ID VARCHAR(255),
    Name VARCHAR(255),
    Other_names VARCHAR(255),
    Wikidata TEXT,
    Mendeley TEXT,
    UniProt TEXT,
    LmjF_ortholog VARCHAR(255),
    LmjFC_ortholog VARCHAR(255),
    LdHU3_ortholog VARCHAR(255),
    LBRM2904_ortholog VARCHAR(255)
    )
    """)

# Insert the data
df_main.to_sql("infantum", conn, if_exists="replace", index=False)

# Check if the data was inserted
cursor.execute("""
               SELECT * 
               FROM infantum 
               LIMIT 10
               """)   
print(cursor.fetchall())

[('LINF_010005000', 'Protein of unknown function (DUF2946)', 'LinJ.01.0010', '<a href="https://www.wikidata.org/wiki/Q62005644" target="_blank">Wikidata</a>', '<a href="https://data.mendeley.com/datasets/m9mggnd5w4/1" target="_blank">Mendeley</a>', '<a href="https://www.uniprot.org/uniprotkb?query=%28gene%3ALINF_010005000%29" target="_blank">UniProt</a>', 'LmjF.01.0010', 'LMJFC_010005100', 'LDHU3_01.0030', None), ('LINF_010005100', 'Endonuclease/Exonuclease/phosphatase family', 'LinJ.01.0020', '<a href="https://www.wikidata.org/wiki/Q65246947" target="_blank">Wikidata</a>', '<a href="https://data.mendeley.com/datasets/d99ycn2bhb/1" target="_blank">Mendeley</a>', '<a href="https://www.uniprot.org/uniprotkb?query=%28gene%3ALINF_010005100%29" target="_blank">UniProt</a>', 'LmjF.01.0020', 'LMJFC_010005200', 'LDHU3_01.0040', 'LBRM2904_01.0010'), ('LINF_010005200', 'Kinesin-13', 'LinJ.01.0030', '<a href="https://www.wikidata.org/wiki/Q65459375" target="_blank">Wikidata</a>', '<a href="https: